### PR TITLE
[automation] Fully synchronize script action/condition execution if supported by engine

### PR DESCRIPTION
**Pull request by florian-h05**
If the `ScriptEngine` implements `Lock`, synchronize the execution of script action/condition.
In contrast to #4402, this now synchronizes the full script context access, which setup of execution context, script evaluation and reset of execution context, and not only the script evaluation.

It moves the locking mechanism added in #4402 from `AbstractScriptModuleHandler` to the inheritors of `AbstractScriptModuleHandler`.

This fixes the bug that https://github.com/openhab/openhab-addons/pull/17510 worked around by using `Thread.sleep`.

### CODE REVIEW

**Comment by florian-h05** [comment]
https://github.com/openhab/openhab-addons/pull/17630 can be merged after this PR here has been merged.

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!

**Pull request merged**
